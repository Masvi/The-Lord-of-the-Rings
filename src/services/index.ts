export { fetchMovies } from './services';
export { fetchQuote } from './services';