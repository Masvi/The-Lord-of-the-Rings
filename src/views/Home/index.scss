.home {
  &__list {
    display: grid;
    grid-template-columns: repeat(1, 1fr);
    margin: 0 auto;
    max-width: 900px;
    gap: 1%;
    padding: 20px;

    &-full-column {
      grid-template-columns: repeat(4, 1fr);
      gap: 1%;
    }
  }

  &__header {
    width: 100%;
    height: 200px;
    grid-column: span 4;
  }
}

/* tablet */
@media (min-width: 700px) {
  .home__list {
    grid-template-columns: repeat(2, 1fr);
  }
}

/* desktop */
@media (min-width: 900px) {
  .home__list {
    grid-template-columns: repeat(4, 1fr);
  }
}
