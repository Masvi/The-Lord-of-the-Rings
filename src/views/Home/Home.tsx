import { ChangeEvent, useEffect, useState } from "react";

import "./index.scss";
import Movie from "../../models/Movie";
import { fetchMovies, fetchQuote } from "../../services";

import { Card, Header, Modal } from "../../components";

import { ModalProps } from "../../components/Modal/Modal";

import { calculateMovieAverages, sortMovies } from '../../utils';

interface Average {
  runtime: number;
  budget: number;
}

const Home = () => {
  const [movieList, setMovieList] = useState<Movie[]>([]);
  const [unfiltered, setUnfiltered] = useState<Movie[]>([]);
  const [average, setAverage] = useState<Average>({ runtime: 0, budget: 0 })
  const [openModal, setOpenModal] = useState<boolean>(false);
  const [movieData, setMovieData] = useState<ModalProps['data']>({});

  const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    if (value) {
      const filtered = movieList.filter((movie) => movie.name
        .toLowerCase()
        .includes(value.toLowerCase()));

      if (filtered.length > 0) {
        setMovieList([...filtered]);
      }
      return;
    }
    setMovieList(unfiltered);
  };

  const handleSort = (e: ChangeEvent<HTMLInputElement>) => {
    const sortedMovies = sortMovies({ movieList, e });
    setMovieList(sortedMovies)
  };

  const handleOpenModal = (movie: Movie) => {
    const { _id } = movie;

    setMovieData({ movie, complete: false })
    setOpenModal(true);

    if (_id) {
      fetchQuote(_id).then((data) => {
        setMovieData({ quote: data, movie, complete: true });
      });
    }
  };

  useEffect(() => {
    fetchMovies().then((data) => {
      setMovieList(data);
      setUnfiltered(data);
      const avareges = calculateMovieAverages(data);
      setAverage(avareges);

    }).catch((e) => console.log(e))
  }, []);

  return (
    <div className="home" data-testid="movie-home">
      <Modal
        data={movieData}
        isOpen={openModal}
        onClose={() => setOpenModal(false)}
      />
      <Header
        handleChange={handleSearchChange}
        handleChangeSort={(e) => handleSort(e)}
        average={average}
      />
      <div className="home__list" data-testid="movie-list">
        {movieList.length > 0 ? (movieList).map((movie) =>
          <Card
            handleClick={() => handleOpenModal(movie)}
            key={movie.name}
            movie={movie}
          />
        ) : <>Loading...</>}
      </div>
    </div>
  )
};

export default Home;