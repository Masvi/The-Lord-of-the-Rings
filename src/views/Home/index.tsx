import { ChangeEvent, useEffect, useState } from "react";

import "./index.scss";
import Movie from "../../models/Movie";
import { fetchMovies, fetchQuote } from "../../services";

import Card from "../../components/Card";
import Modal from "../../components/Modal";
import Header from "../../components/Header";

import { ModalProps } from "../../components/Modal";

interface Average {
  runtime: number;
  budget: number;
}

const Home = () => {
  const [movieList, setMovieList] = useState<Movie[]>([]);
  const [unfiltered, setUnfiltered] = useState<Movie[]>([]);
  const [average, setAverage] = useState<Average>({ runtime: 0, budget: 0 })
  const [openModal, setOpenModal] = useState<boolean>(false);
  const [movieData, setMovieData] = useState<ModalProps['data']>({});

  const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    if (value) {
      const filtered = movieList.filter((movie) => movie.name
        .toLowerCase()
        .includes(value.toLowerCase()));

      if (filtered.length > 0) {
        setMovieList([...filtered]);
      }
      return;
    }
    setMovieList(unfiltered);
  };

  const handleChangeSort = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    const sortedMovies = movieList.sort((first, second) => {
      const movieA = first.name.toLowerCase();
      const movieB = second.name.toLowerCase();

      const firstMovie = value === 'asc' ? movieA : movieB
      const secondMovie = value === 'asc' ? movieB : movieA

      if (firstMovie > secondMovie) {
        return 1
      } else if (firstMovie < secondMovie) {
        return -1;
      } else {
        return 0;
      }
    })
    setMovieList([...sortedMovies])
  };

  const handleAverage = (data: Movie[]) => {
    if (data.length > 0) {
      let runtime = data.reduce(
        (acc, movie) => acc + movie.runtimeInMinutes,
        0
      );
      let budget = data.reduce(
        (acc, movie) => acc + movie.budgetInMillions,
        0
      );
      runtime = runtime / data?.length
      budget = budget / data?.length
      setAverage({ runtime, budget });
    }
  };

  const handleOpenModal = (movie: Movie) => {
    const { _id } = movie;

    setMovieData({ movie, complete: false })
    setOpenModal(true);

    if (_id) {
      fetchQuote(_id).then((data) => {
        setMovieData({ quote: data, movie, complete: true });
      });
    }
  };

  useEffect(() => {
    fetchMovies().then((data) => {
      setMovieList(data);
      setUnfiltered(data);
      handleAverage(data);
    }).catch((e) => console.log(e))
  }, []);

  return (
    <div className="home">
      <Modal
        data={movieData}
        isOpen={openModal}
        onClose={() => setOpenModal(false)}
      />
      <Header
        handleChange={handleSearchChange}
        handleChangeSort={handleChangeSort}
        average={average}
      />
      <div className="home__list">
        {movieList.length > 0 ? (movieList).map((movie) =>
          <Card
            handleClick={() => handleOpenModal(movie)}
            key={movie.name}
            movie={movie}
          />
        ) : <>Loading...</>}
      </div>
    </div>
  )
};

export default Home;