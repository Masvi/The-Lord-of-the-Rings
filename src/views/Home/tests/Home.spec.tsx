
import { render, screen, within } from "@testing-library/react"
import { fetchMovies } from "../../../services";

import Home from "../Home";
import userEvent from "@testing-library/user-event";

jest.mock('axios');

const mockData = {
  docs: [
    {
      _id: 'movie_id_1',
      name: 'movieName',
      academyAwardNominations: 10,
      academyAwardWins: 2,
      boxOfficeRevenueInMillions: 290,
      budgetInMillions: 291,
      rottenTomatoesScore: 94,
      runtimeInMinutes: 559,
    },
    {
      _id: 'movie_id_2',
      name: 'movieName 2',
      academyAwardNominations: 11,
      academyAwardWins: 2,
      boxOfficeRevenueInMillions: 290,
      budgetInMillions: 391,
      rottenTomatoesScore: 94,
      runtimeInMinutes: 659,
    },
    {
      _id: 'movie_id_3',
      name: 'movieName 3',
      academyAwardNominations: 11,
      academyAwardWins: 2,
      boxOfficeRevenueInMillions: 190,
      budgetInMillions: 191,
      rottenTomatoesScore: 94,
      runtimeInMinutes: 600,
    }
  ],
}

jest.mock('../../../services', () => ({
  fetchMovies: jest.fn(),
}));

describe("Home", () => {
  test("should render correctly", async () => {
    (fetchMovies as jest.Mock).mockResolvedValue([]);
    const { container } = render(<Home />);

    expect(await screen.findByTestId('movie-home')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("should display header and loading correctly for empty data", async () => {
    (fetchMovies as jest.Mock).mockResolvedValue([]);
    render(<Home />);

    expect(await screen.findByText(/The Lord of the Rings movies/i)).toBeInTheDocument();
    expect(await screen.findByText(/Loading.../i)).toBeInTheDocument();
  });

  test("should display list element correctly", async () => {
    (fetchMovies as jest.Mock).mockResolvedValue([mockData.docs[0]]);
    render(<Home />);

    expect(await screen.findByText(/movieName/i)).toBeInTheDocument();
    expect(await screen.findByText(/2 wins and 10 nominations/i)).toBeInTheDocument();
    expect(await screen.findByText(/291M/i)).toBeInTheDocument();
    expect(await screen.findByText(/290M/i)).toBeInTheDocument();
  });

  test("should return avarage values correctly", async () => {
    (fetchMovies as jest.Mock).mockResolvedValue([...mockData.docs]);
    render(<Home />);

    expect(await screen.findByText(/Average movie runtime: 606 min/i)).toBeInTheDocument();
    expect(await screen.findByText(/Average movie budget: 291 M/i)).toBeInTheDocument();
  });

  test("should sort elements values correctly", async () => {
    (fetchMovies as jest.Mock).mockResolvedValue([...mockData.docs]);
    render(<Home />);

    const selectElement = screen.getByRole('combobox');
    const optionToSelect = screen.getByRole('option', { name: 'Sort by Desc' });
    await userEvent.selectOptions(selectElement, optionToSelect);
    const movieListElement = await screen.findAllByTestId('card-title');

    expect(movieListElement).toHaveLength(3);
    expect(movieListElement.map(title => title.textContent)).toEqual(['movieName 3', 'movieName 2', 'movieName']);
  });
});