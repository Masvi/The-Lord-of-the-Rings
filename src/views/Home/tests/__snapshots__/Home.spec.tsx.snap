// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div>
  <div
    class="home"
    data-testid="movie-home"
  >
    <div
      class="header"
    >
      <div
        class="header__ave-infos"
      >
        <h1>
          The Lord of the Rings movies
        </h1>
        <p>
          Average movie runtime: 
          0
           min
        </p>
        <p>
          Average movie budget: 
          0
           M
        </p>
      </div>
      <div
        class="header__inputs"
      >
        <input
          class="search-input"
          placeholder="search by name"
          type="search"
        />
        <select
          class="select-sort"
        >
          <option
            value="asc"
          >
            Sort by Asc
          </option>
          <option
            value="desc"
          >
            Sort by Desc
          </option>
        </select>
      </div>
    </div>
    <div
      class="home__list"
      data-testid="movie-list"
    >
      Loading...
    </div>
  </div>
</div>
`;
